Some LightMetal capture + replay support for initial TTNN tests/models - Round 1 (pybind, sharded buffers, bug fix)

kmabeeTT opened the pull request:
### Ticket

https://github.com/tenstorrent/tt-metal/issues/18145

### Problem description

Add more support for things encountered/needed by TTNN w/ Light Metal capture + replay. This won't close ticket above, but is step in that direction. More to come over coming days/weeks.

### What's changed

Split into 3 seperate commits for easier viewing.

- Expose/ pybind LightMetalBegin/EndCapture() APIs as ttnn.light_metal_begin/end_capture() - I use them in some WIP tests not included in this PR (tests need more work).
- Move trace/replay of CreateProgram() to Program() constructor and CreateBuffer() to `Buffer::create()` (now includes sharded buffer support) after discussion offline, this negates the need for https://github.com/tenstorrent/tt-metal/pull/17781 
- Handle optionals in `CreateCircularBuffer()` path safer which fixes a corner case bug expose by rn50 tracing

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes

kmabeeTT committed:
LightMetal - pybind LightMetalBegin/EndCapture() APIs as ttnn.light_metal_begin/end_capture() (Issue #18145)

kmabeeTT committed:
LightMetal - Capture/Replay support for Program() and Buffer::create() to replace host apis (Issue #18145)

 - Negates the need for a previous PR I put up to switch TTNN ops to use
   CreateBuffer() and CreateProgram() in few places.
 - Don't trace CreateProgram() or CreateBuffer() APIs anymore
 - Now supports sharded buffers too for first time, needed by rn50,
   though I copied some logic from TTNN tensor
   serialization/deserialization here, eventually consolidate.
 - Split up CoreSpec variant to/from_flatbuffer() functions for
   CoreCoord, CoreRange, CoreRangeSet for reuse by ShardSpec's CoreRangeSet

 - Tests: Add some tt-metal unit tests for interleaved/sharded buffers
   and using Buffer::create() and Program() which are now supported,
   also add some initial simple python ttnn tests with light metal capture.
   Needs more work to be able to support reply in test, and comparison
   between capture and replay, that will come in future.

kmabeeTT committed:
LightMetal - Handle optionals in CreateCircularBuffer() capture/replay safer (Issue #18145)

 - Don't rely on value of 0 denoting globally_allocated_addr or shadow
   buffer of global_id=0 not existing. Use optional tables. Resnet was
   having shadow buffer with global_id=0 which hit issue here.

kmabeeTT commented:
Whoops, did a rebase before pushing to this branch, turns out there is an update at 2d4f9945fb to `ShardSpecBuffer` I'll have to make some corresponding tweaks to for serialization/deserialization will force push once fixed. Other changes are probably unaffected.

omilyutin-tt commented on ttnn/cpp/pybind11/core.hpp:
```diff
@@ -30,7 +31,20 @@ void py_module(py::module& module) {
     });
     py_config.def_property_readonly("report_path", &ttnn::Config::get<"report_path">);
 
+    py::class_<LightMetalBinary>(module, "LightMetalBinary")
+        .def(py::init<>())
+        .def(py::init<std::vector<uint8_t>>())
+        .def("get_data", &LightMetalBinary::get_data)
+        .def("set_data", &LightMetalBinary::set_data)
+        .def("size", &LightMetalBinary::size)
+        .def("is_empty", &LightMetalBinary::is_empty)
+        .def("save_to_file", &LightMetalBinary::save_to_file)
+        .def_static("load_from_file", &LightMetalBinary::load_from_file);
+
     module.def("get_memory_config", &ttnn::get_memory_config);
+    module.def("light_metal_begin_capture", &LightMetalBeginCapture);
+    module.def("light_metal_end_capture", &LightMetalEndCapture);
```
A separate file for this might be a good idea

kmabeeTT committed:
LightMetal - Add and use LightMetalReplay::clear_object_maps() to clear state after binary run (Issue #18145)

 - Prevents "Attempting to push work to Device 0 which is not initialized. Ignoring..." fatal
   due to CloseDevice() currently called explicitly by replay before replay context
   destroyed which may contain Programs with Buffers not yet deallocated

cfjchu commented:
Overall, looks good but if we're adding instrumentation to tt-metal, my expectation is we have tt-metal level tests to exercise this behavior. Otherwise we're relying on TT-Forge to excercise and regress on this behavior.

kmabeeTT commented:
> Overall, looks good but if we're adding instrumentation to tt-metal, my expectation is we have tt-metal level tests to exercise this behavior. Otherwise we're relying on TT-Forge to excercise and regress on this behavior.

Thanks @cfjchu that is fair,  I added some basic tests and squashed into commit 92baa60000 just now (tt-metal and ttnn tests, including using Program constructor and Buffer::create() in tt-metal tests, and some light testing on sharded buffers, and pair of simple ttnn single and multi-op light metal tracing tests).

cfjchu commented on tests/ttnn/unit_tests/test_single_device_light_metal_trace.py:
```diff
@@ -0,0 +1,68 @@
+    # logger.info("ttnn_torch_output_tensor: {}".format(ttnn_torch_output_tensor))
+
+    binary_data = ttnn.light_metal_end_capture()
+    binary_data.save_to_file(f"/tmp/test_single_op_test_light_metal_capture.bin")
```
Use `tmp_path` fixture: https://docs.pytest.org/en/stable/how-to/tmp_path.html

kmabeeTT replied:
Thanks, I made update and included a helper function to drop a line about path and file size used.

```
2025-02-26 01:54:14.749 | INFO     | tests.ttnn.unit_tests.test_single_device_light_metal_trace:write_binary_to_file:18 - Writing light metal binary 10008 bytes to /tmp/pytest-of-kmabee/pytest-12/test_single_op_test_light_meta0/test_single_op_test_light_metal_capture.bin
2025-02-26 01:54:14.760 | INFO     | tests.ttnn.unit_tests.test_single_device_light_metal_trace:write_binary_to_file:18 - Writing light metal binary 10008 bytes to /tmp/pytest-of-kmabee/pytest-12/test_single_op_test_light_meta1/test_single_op_test_light_metal_capture.bin
2025-02-26 01:54:14.766 | INFO     | tests.ttnn.unit_tests.test_single_device_light_metal_trace:write_binary_to_file:18 - Writing light metal binary 9796 bytes to /tmp/pytest-of-kmabee/pytest-12/test_single_op_test_light_meta2/test_single_op_test_light_metal_capture.bin
2025-02-26 01:54:15.941 | INFO     | tests.ttnn.unit_tests.test_single_device_light_metal_trace:write_binary_to_file:18 - Writing light metal binary 48268 bytes to /tmp/pytest-of-kmabee/pytest-12/test_chain_op_test_light_metal0/test_chain_op_test_light_metal_capture.bin
2025-02-26 01:54:16.815 | INFO     | tests.ttnn.unit_tests.test_single_device_light_metal_trace:write_binary_to_file:18 - Writing light metal binary 48268 bytes to /tmp/pytest-of-kmabee/pytest-12/test_chain_op_test_light_metal1/test_chain_op_test_light_metal_capture.bin
2025-02-26 01:54:17.638 | INFO     | tests.ttnn.unit_tests.test_single_device_light_metal_trace:write_binary_to_file:18 - Writing light metal binary 24620 bytes to /tmp/pytest-of-kmabee/pytest-12/test_chain_op_test_light_metal2/test_chain_op_test_light_metal_capture.bin
```

Force pushed, squashed, marking resolved.

cfjchu commented on tests/ttnn/unit_tests/test_single_device_light_metal_trace.py:
```diff
@@ -0,0 +1,68 @@
+@pytest.mark.parametrize("blocking", [True])
+@pytest.mark.parametrize("device_params", [{"trace_region_size": 200000}])
+def test_chain_op_test_light_metal_capture(device, shape, enable_async, blocking):
+    ttnn.light_metal_begin_capture()
+
+    device.enable_async(enable_async)
+    device.enable_program_cache()
+    input_0_dev = ttnn.allocate_tensor_on_device(ttnn.Shape(shape), ttnn.bfloat16, ttnn.TILE_LAYOUT, device)
+    input_1_dev = ttnn.allocate_tensor_on_device(ttnn.Shape(shape), ttnn.bfloat16, ttnn.TILE_LAYOUT, device)
+
+    # Op chain to be traced
+    def run_op_chain(input_0, input_1):
+        return ttnn.neg(ttnn.add(ttnn.mul(input_1, ttnn.neg(ttnn.gelu(input_0))), ttnn.relu(input_1)))
+
+    output_tensor = run_op_chain(input_0_dev, input_1_dev)
+
+    ttnn_torch_output_tensor = ttnn.to_torch(output_tensor)
+    # logger.info("ttnn_torch_output_tensor: {}".format(ttnn_torch_output_tensor))
+
+    binary_data = ttnn.light_metal_end_capture()
+    binary_data.save_to_file(f"/tmp/test_chain_op_test_light_metal_capture.bin")
```
very cool to see end to end light metal trace capture work. Does replay not yet work?

kmabeeTT replied:
It's in tt-metal tests, but it will take more thinking and figuring out how to make it work with TTNN. Will be circling back to this and other things next as focus on TTNN increases.
